### PR TITLE
feat(pmp-api): fetch data from bitbucket

### LINKED ISSUE #7: Fetch data about PRs from bitbucket

### PR DESCRIPTION
**Pull request by wjanaszek**
Integrate with bitbucket.

Resolves #7

### CODE REVIEW

File: libs/server/repository/infrastructure/src/lib/bitbucket/mappers/map-bitbucket-pr-details.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+import { BitbucketPrDetailsEntity } from '../domain/entities/bitbucket-pr-details.entity';
+import { mapBitbucketAccount } from './map-bitbucket-account';
+
+export const mapBitbucketPrDetails = (pr: BitbucketPrDetailsEntity): PrEntity =>
+  plainToClass(PrEntity, {
+    ...pr,
+    additions: pr.diff.map(diff => diff.lines_added).reduce((prev, curr) => prev + curr),
+    changedFiles: pr.diff.length,
+    createdAt: pr.created_on,
+    deletions: pr.diff.map(diff => diff.lines_removed).reduce((prev, curr) => prev + curr),
```
Comment (Sikora00) [inline libs/server/repository/infrastructure/src/lib/bitbucket/mappers/map-bitbucket-pr-details.ts]: maybe calculate deletions and additions on the beginning inside one iteration through the diff array.

File: libs/server/repository/infrastructure/src/lib/bitbucket/repositories/bitbucket-repository.repository.ts [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+import { catchError, map } from 'rxjs/operators';
+import { catchRequestExceptions } from '@pimp-my-pr/server/shared/util-exception';
+import { CoreException, CoreNotFoundException } from '@pimp-my-pr/server/shared/domain';
+import { throwError } from 'rxjs';
+import { AxiosError, AxiosResponse } from 'axios';
+import { mapBitbucketRepository } from '../mappers/map-bitbucket-repository';
+import { BitbucketRepositoryEnity } from '../domain/entities/bitbucket-repository.enity';
+
+@Injectable()
+export class GithubRepositoryRepository extends RepositoryRepository {
```
Comment (Sikora00) [inline libs/server/repository/infrastructure/src/lib/bitbucket/repositories/bitbucket-repository.repository.ts]: Name copy paste

File: libs/server/shared/core/src/lib/config/.env-sample [hunk 3]
```diff
@@ -1,2 +1,6 @@
+// bitbucket
+PMP_API_SERVICE_BITBUCKET_TOKEN=
+
+// github
 PMP_API_SERVICE_GITHUB_TOKEN=
-PMP_API_SERVICE_REPOSITORIES=valueadd-poland/pimp-my-pr,angular/angular
+PMP_API_SERVICE_REPOSITORIES=
```
Comment (Sikora00) [inline libs/server/shared/core/src/lib/config/.env-sample]: maybe consider leaving it as it was. I wanted an example of what is the format of this variable.

File: libs/server/repository/infrastructure/src/lib/bitbucket/repositories/bitbucket-reviewer.repository.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+import { urlFactory } from '@valueadd/typed-urls';
+import { bitbucketConfig } from '@pimp-my-pr/server/shared/core';
+import { ReviewerEntity } from '@pimp-my-pr/server/repository/core/domain';
+import { map } from 'rxjs/operators';
+import { catchRequestExceptions } from '@pimp-my-pr/server/shared/util-exception';
+import { BitbucketUserEntity } from '../domain/entities/bitbucket-user.entity';
+import { mapBitbucketUser } from '../mappers/map-bitbucket-user';
+
+@Injectable()
+export class GithubReviewerRepository extends ReviewerRepository {
```
Comment (stefanoww) [inline libs/server/repository/infrastructure/src/lib/bitbucket/repositories/bitbucket-reviewer.repository.ts]: Shouldn't be BitbucketReviewe?

File: libs/server/shared/core/src/lib/server-shared-core.module.ts [hunk 5]
```diff
@@ -1,23 +1,38 @@
...
 
 @Global()
 @Module({
   imports: [ConfigModule, HttpModule],
-  providers: [GithubAuthInterceptor],
-  exports: [ConfigModule, HttpModule, GithubAuthInterceptor]
+  providers: [
+    {
+      provide: AuthInterceptor,
+      useFactory: (configService: PmpApiServiceConfigService, httpService: HttpService) => {
```
Comment (stefanoww) [inline libs/server/shared/core/src/lib/server-shared-core.module.ts]: Maybe move it to the factory method

File: libs/server/repository/core/domain/src/lib/entities/pr.entity.ts [hunk 6]
```diff
@@ -20,7 +20,7 @@ export class PrEntity {
   @f
   id: number;
   @f
-  reviewCommentsCount: number;
+  reviewCommentsCount?: number;
```
Comment (stefanoww) [inline libs/server/repository/core/domain/src/lib/entities/pr.entity.ts]: Shouldn't we use @f.optional() in this case?

File: libs/server/repository/core/application-services/src/lib/read-models/interfaces/pr-author.interface.ts [hunk 7]
```diff
@@ -0,0 +1,5 @@
+export interface PrAuthor {
```
Comment (Sikora00) [inline libs/server/repository/core/application-services/src/lib/read-models/interfaces/pr-author.interface.ts]: It was changed to the AuthorEntity

File: libs/server/repository/core/application-services/src/lib/read-models/interfaces/pr-reviewer.interface.ts [hunk 8]
```diff
@@ -0,0 +1,5 @@
+export interface PrReviewer {
```
Comment (Sikora00) [inline libs/server/repository/core/application-services/src/lib/read-models/interfaces/pr-reviewer.interface.ts]: It was changed to the ReviewerEntity

File: libs/server/repository/core/domain/src/lib/entities/pr.entity.ts [hunk 9]
```diff
@@ -10,16 +10,16 @@ export class PrEntity {
...
   commentsCount: number;
   @f
   createdAt: Date;
   @f
   deletions: number;
   @f
-  id: number;
-  @f
+  id: string;
+  @f.optional()
```
Comment (Sikora00) [inline libs/server/repository/core/domain/src/lib/entities/pr.entity.ts]: if this is optional then property also should be

File: libs/server/repository/core/domain/src/lib/entities/pr.entity.ts [hunk 10]
```diff
@@ -10,16 +10,16 @@ export class PrEntity {
   @f
   changedFiles: number;
   @f
-  closedAt: Date;
+  closedAt?: Date;
```
Comment (Sikora00) [inline libs/server/repository/core/domain/src/lib/entities/pr.entity.ts]: if this is optional then @f.optional() also should be

File: libs/server/repository/infrastructure/src/index.ts [hunk 11]
```diff
@@ -1,4 +1,9 @@
 export * from './lib/server-repository-infrastructure.module';
+
+export * from './lib/bitbucket/repositories/bitbucket-pr.repository';
```
Comment (Sikora00) [inline libs/server/repository/infrastructure/src/index.ts]: do you need to export that?

File: libs/server/repository/infrastructure/src/lib/bitbucket/repositories/bitbucket-repository.repository.ts [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+            return throwError(new RepositoryNotFoundException(fullName));
+          }
+          return throwError(error);
+        })
+      )
+      .toPromise();
+  }
+
+  // @TODO pass workspace dynamically
+  getSingleRepository(id: string, workspace = 'valueadd'): Promise<RepositoryEntity> {
```
Comment (Sikora00) [inline libs/server/repository/infrastructure/src/lib/bitbucket/repositories/bitbucket-repository.repository.ts]: what happend here?

File: libs/server/repository/infrastructure/src/lib/repositories/repository-repository.adapter.ts [hunk 13]
```diff
@@ -3,13 +3,12 @@ import { InjectRepository } from '@nestjs/typeorm';
...
 import { RepositoryRepository } from '@pimp-my-pr/server/repository/core/domain-services';
 import { Repository } from 'typeorm';
-import { GithubRepositoryRepository } from '../github/repositories/github-repository.repository';
 
 @Injectable()
 export class RepositoryRepositoryAdapter extends RepositoryRepository {
   constructor(
     @InjectRepository(RepositoryEntity) private typeOrmRepository: Repository<RepositoryEntity>,
-    private githubRepository: GithubRepositoryRepository
+    private repository: RepositoryRepository
```
Comment (Sikora00) [inline libs/server/repository/infrastructure/src/lib/repositories/repository-repository.adapter.ts]: RepositoryRepositoryAdapter extends RepositoryRepository. I think you meant something like `remoteRepository: RemoteRepositoryRepository`

File: libs/server/repository/shell/src/lib/factories/repository-repository.factory.ts [hunk 14]
```diff
@@ -0,0 +1,20 @@
+import { PmpApiConfigService } from '@pimp-my-pr/server/shared/core';
+import { HttpService } from '@nestjs/common';
+import {
+  BitbucketRepositoryRepository,
+  GithubRepositoryRepository
+} from '@pimp-my-pr/server/repository/infrastructure';
+
+export const repositoryRepositoryFactory = (
```
Comment (Sikora00) [inline libs/server/repository/shell/src/lib/factories/repository-repository.factory.ts]: remoteRepositoryRepositoryFactory

**Pull request merged**